In [48]:
# Problem 4: Matrix Transpose and Inverse
# Problem 4 Part 2
import numpy as np

# generate 3 3x3 matrices 
M_1 = np.random.randint(-10, 11, size = (3,3))
M_2 = np.random.randint(-10, 11, size = (3,3))
M_3 = np.random.randint(-10, 11, size = (3,3))

print("Matrix 1: ", M_1)
print("Matrix 2: ", M_2)
print("Matrix 3: ", M_3)

Matrix 1:  [[-10   1  -3]
 [  6   0   4]
 [ -3   0  -4]]
Matrix 2:  [[-6  2 -3]
 [ 7 -1 10]
 [ 3 -8  3]]
Matrix 3:  [[-6 -3  3]
 [ 8 10  1]
 [-2  5  7]]


In [60]:
# calculate inverses using packages
M_1_inv = np.linalg.inv(M_1)
M_2_inv = np.linalg.inv(M_2)
M_3_inv = np.linalg.inv(M_3)

# round to two decimals for readability 
print("Matrix 1 Inverse: ", np.round(M_1_inv, decimals = 2))
print("Matrix 2 Inverse: ", np.round(M_2_inv, decimals = 2))
print("Matrix 3 Inverse: ", np.round(M_3_inv, decimals = 2))

Matrix 1 Inverse:  [[-0.    0.33  0.33]
 [ 1.    2.58  1.83]
 [-0.   -0.25 -0.5 ]]
Matrix 2 Inverse:  [[-0.27 -0.06 -0.06]
 [-0.03  0.03 -0.14]
 [ 0.19  0.15  0.03]]
Matrix 3 Inverse:  [[-1.81 -1.    0.92]
 [ 1.61  1.   -0.83]
 [-1.67 -1.    1.  ]]


In [62]:
# compute product of matrix and inverse to check for identity matrix
prod_1 = M_1 @ M_1_inv
prod_2 = M_2 @ M_2_inv
prod_3 = M_3 @ M_3_inv

# round to two decimals for readability 
print("Matrix 1 and Inverse Product: ", np.round(prod_1, decimals = 2))
print("Matrix 2 and Inverse Product: ", np.round(prod_2, decimals = 2))
print("Matrix 3 and Inverse Product: ", np.round(prod_3, decimals = 2))

Matrix 1 and Inverse Product:  [[ 1.  0. -0.]
 [ 0.  1.  0.]
 [ 0.  0.  1.]]
Matrix 2 and Inverse Product:  [[ 1. -0. -0.]
 [-0.  1. -0.]
 [-0. -0.  1.]]
Matrix 3 and Inverse Product:  [[ 1.  0. -0.]
 [ 0.  1.  0.]
 [-0.  0.  1.]]


In [ ]:
# Problem 5: Average, variance, and correlation

import csv
import pandas as pd
house_data = pd.read_csv("kc_house_data.csv")
house_data = house_data.drop(columns = ["id", "date", "zipcode"])

In [52]:
# For each feature, write code to compute the average value, the min and max values, as well as its variance
stats = house_data.agg(["mean", "min", "max", "var"])
print(stats)

             price   bedrooms  bathrooms    sqft_living      sqft_lot  \
mean  5.400881e+05   3.370842   2.114757    2079.899736  1.510697e+04   
min   7.500000e+04   0.000000   0.000000     290.000000  5.200000e+02   
max   7.700000e+06  33.000000   8.000000   13540.000000  1.651359e+06   
var   1.347824e+11   0.865015   0.593151  843533.681368  1.715659e+09   

        floors  waterfront      view  condition      grade     sqft_above  \
mean  1.494309    0.007542  0.234303   3.409430   7.656873    1788.390691   
min   1.000000    0.000000  0.000000   1.000000   1.000000     290.000000   
max   3.500000    1.000000  4.000000   5.000000  13.000000    9410.000000   
var   0.291588    0.007485  0.587243   0.423467   1.381703  685734.667269   

      sqft_basement     yr_built   yr_renovated        lat        long  \
mean     291.509045  1971.005136      84.402258  47.560053 -122.213896   
min        0.000000  1900.000000       0.000000  47.155900 -122.519000   
max     4820.000000  2015.

In [63]:
# answering the questions for question 1 regarding the mean and the variance
high_mean = stats.loc["mean"].max()
high_mean_feature = stats.loc["mean"].idxmax()
print("The feature with the highest mean is", high_mean_feature, high_mean)

The feature with the highest mean is price 540088.1417665294


- Feature with the highest average: sqft_living 2079.899736
- Feature with the lowest average: waterfront 0.007542
- Feature with highest variance:
- Feature with the lowest variance: 